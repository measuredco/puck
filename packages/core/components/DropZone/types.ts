import { CSSProperties } from "react";

export type DropZoneProps = {
  zone: string;
  allow?: string[];
  disallow?: string[];
  style?: CSSProperties;
};
