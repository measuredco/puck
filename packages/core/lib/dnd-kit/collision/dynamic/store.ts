import { createStore } from "zustand/vanilla";

export const collisionStore = createStore<{
  fallbackEnabled: boolean;
}>(() => ({
  fallbackEnabled: false,
}));
