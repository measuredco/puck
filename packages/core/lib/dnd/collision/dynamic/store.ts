import { createStore } from "zustand/vanilla";
import { Direction } from "../../../../types";

export type CollisionMap = Record<
  string,
  {
    direction: Direction;
    [key: string]: any;
  }
>;

export const collisionStore = createStore<{
  fallbackEnabled: boolean;
  direction: Direction;
  collisionMap: CollisionMap;
}>(() => ({
  fallbackEnabled: false,
  direction: null,
  collisionMap: {},
}));
