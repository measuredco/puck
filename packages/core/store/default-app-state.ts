import { defaultViewports } from "../components/ViewportControls/default-viewports";
import { PrivateAppState } from "../types/Internal";

export const defaultAppState: PrivateAppState = {
  data: { content: [], root: {}, zones: {} },
  ui: {
    leftSideBarVisible: true,
    rightSideBarVisible: true,
    arrayState: {},
    itemSelector: null,
    componentList: {},
    isDragging: false,
    previewMode: "edit",
    viewports: {
      current: {
        width: defaultViewports[0].width,
        height: defaultViewports[0].height || "auto",
      },
      options: [],
      controlsVisible: true,
    },
    field: { focus: null },
  },
  indexes: {
    nodes: {},
    zones: {},
  },
};
