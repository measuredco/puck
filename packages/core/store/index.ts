import {
  Config,
  IframeConfig,
  Overrides,
  AppState,
  UiState,
  Plugin,
  UserGenerics,
  Field,
  ComponentConfig,
  Metadata,
  ComponentData,
  RootDataWithProps,
  ResolveDataTrigger,
} from "../types";
import { createReducer, PuckAction } from "../reducer";
import { getItem } from "../lib/data/get-item";
import { defaultViewports } from "../components/ViewportControls/default-viewports";
import { Viewports } from "../types";
import { create, StoreApi, useStore } from "zustand";
import { subscribeWithSelector } from "zustand/middleware";
import { createContext, useContext } from "react";
import { createHistorySlice, type HistorySlice } from "./slices/history";
import { createNodesSlice, type NodesSlice } from "./slices/nodes";
import {
  createPermissionsSlice,
  type PermissionsSlice,
} from "./slices/permissions";
import { createFieldsSlice, type FieldsSlice } from "./slices/fields";
import { PrivateAppState } from "../types/Internal";
import { resolveComponentData } from "../lib/resolve-component-data";
import { walkTree } from "../lib/data/walk-tree";
import { toRoot } from "../lib/data/to-root";
import { generateId } from "../lib/generate-id";

export const defaultAppState: PrivateAppState = {
  data: { content: [], root: {}, zones: {} },
  ui: {
    leftSideBarVisible: true,
    rightSideBarVisible: true,
    arrayState: {},
    itemSelector: null,
    componentList: {},
    isDragging: false,
    previewMode: "edit",
    viewports: {
      current: {
        width: defaultViewports[0].width,
        height: defaultViewports[0].height || "auto",
      },
      options: [],
      controlsVisible: true,
    },
    field: { focus: null },
  },
  indexes: {
    nodes: {},
    zones: {},
  },
};

export type Status = "LOADING" | "MOUNTED" | "READY";

type ZoomConfig = {
  autoZoom: number;
  rootHeight: number;
  zoom: number;
};

type ComponentState = Record<string, { loadingCount: number }>;

export type AppStore<
  UserConfig extends Config = Config,
  G extends UserGenerics<UserConfig> = UserGenerics<UserConfig>
> = {
  state: G["UserAppState"];
  dispatch: (action: PuckAction) => void;
  config: UserConfig;
  componentState: ComponentState;
  setComponentState: (componentState: ComponentState) => void;
  setComponentLoading: (
    id: string,
    loading?: boolean,
    defer?: number
  ) => () => void;
  unsetComponentLoading: (id: string) => void;
  pendingLoadTimeouts: Record<string, NodeJS.Timeout>;
  resolveComponentData: <T extends ComponentData | RootDataWithProps>(
    componentData: T,
    trigger: ResolveDataTrigger
  ) => Promise<{ node: T; didChange: boolean }>;
  resolveAndCommitData: () => void;
  plugins: Plugin[];
  overrides: Partial<Overrides>;
  viewports: Viewports;
  zoomConfig: ZoomConfig;
  setZoomConfig: (zoomConfig: ZoomConfig) => void;
  status: Status;
  setStatus: (status: Status) => void;
  iframe: IframeConfig;
  selectedItem?: G["UserData"]["content"][0] | null;
  setUi: (ui: Partial<UiState>, recordHistory?: boolean) => void;
  getComponentConfig: (type?: string) => ComponentConfig | null | undefined;
  onAction?: (action: PuckAction, newState: AppState, state: AppState) => void;
  metadata: Metadata;
  fields: FieldsSlice;
  history: HistorySlice;
  nodes: NodesSlice;
  permissions: PermissionsSlice;
};

export type AppStoreApi = StoreApi<AppStore>;

const defaultPageFields: Record<string, Field> = {
  title: { type: "text" },
};

export const createAppStore = (initialAppStore?: Partial<AppStore>) =>
  create<AppStore>()(
    subscribeWithSelector((set, get) => ({
      state: defaultAppState,
      config: { components: {} },
      componentState: {},
      plugins: [],
      overrides: {},
      viewports: defaultViewports,
      zoomConfig: {
        autoZoom: 1,
        rootHeight: 0,
        zoom: 1,
      },
      status: "LOADING",
      iframe: {},
      metadata: {},
      ...initialAppStore,
      fields: createFieldsSlice(set, get),
      history: createHistorySlice(set, get),
      nodes: createNodesSlice(set, get),
      permissions: createPermissionsSlice(set, get),
      getComponentConfig: (type?: string) => {
        const { config, selectedItem } = get();
        const rootFields = config.root?.fields || defaultPageFields;

        return type && type !== "root"
          ? config.components[type]
          : selectedItem
          ? config.components[selectedItem.type]
          : ({ ...config.root, fields: rootFields } as ComponentConfig);
      },
      selectedItem: initialAppStore?.state?.ui.itemSelector
        ? getItem(
            initialAppStore?.state?.ui.itemSelector,
            initialAppStore.state
          )
        : null,
      dispatch: (action: PuckAction) =>
        set((s) => {
          const { record } = get().history;

          const dispatch = createReducer({
            record,
            appStore: s,
          });

          const state = dispatch(s.state, action);

          const selectedItem = state.ui.itemSelector
            ? getItem(state.ui.itemSelector, state)
            : null;

          get().onAction?.(action, state, get().state);

          return { ...s, state, selectedItem };
        }),
      setZoomConfig: (zoomConfig) => set({ zoomConfig }),
      setStatus: (status) => set({ status }),
      setComponentState: (componentState) => set({ componentState }),
      pendingLoadTimeouts: {},
      setComponentLoading: (
        id: string,
        loading: boolean = true,
        defer: number = 0
      ) => {
        const { setComponentState, pendingLoadTimeouts } = get();

        const loadId = generateId();

        const setLoading = () => {
          const { componentState } = get();

          setComponentState({
            ...componentState,
            [id]: {
              ...componentState[id],
              loadingCount: (componentState[id]?.loadingCount || 0) + 1,
            },
          });
        };

        const unsetLoading = () => {
          const { componentState } = get();

          clearTimeout(timeout);

          delete pendingLoadTimeouts[loadId];

          set({ pendingLoadTimeouts });

          setComponentState({
            ...componentState,
            [id]: {
              ...componentState[id],
              loadingCount: Math.max(
                (componentState[id]?.loadingCount || 0) - 1,
                0
              ),
            },
          });
        };

        const timeout = setTimeout(() => {
          if (loading) {
            setLoading();
          } else {
            unsetLoading();
          }

          delete pendingLoadTimeouts[loadId];

          set({ pendingLoadTimeouts });
        }, defer);

        set({
          pendingLoadTimeouts: {
            ...pendingLoadTimeouts,
            [id]: timeout,
          },
        });

        return unsetLoading;
      },
      unsetComponentLoading: (id: string) => {
        const { setComponentLoading } = get();

        setComponentLoading(id, false);
      },
      // Helper
      setUi: (ui: Partial<UiState>, recordHistory?: boolean) =>
        set((s) => {
          const dispatch = createReducer({
            record: () => {},
            appStore: s,
          });

          const state = dispatch(s.state, {
            type: "setUi",
            ui,
            recordHistory,
          });

          const selectedItem = state.ui.itemSelector
            ? getItem(state.ui.itemSelector, state)
            : null;

          return { ...s, state, selectedItem };
        }),
      resolveComponentData: async (componentData, trigger) => {
        const { config, metadata, setComponentLoading, permissions } = get();

        const timeouts: Record<string, () => void> = {};

        return await resolveComponentData(
          componentData,
          config,
          metadata,
          (item) => {
            const id = "id" in item.props ? item.props.id : "root";
            timeouts[id] = setComponentLoading(id, true, 50);
          },
          async (item) => {
            const id = "id" in item.props ? item.props.id : "root";

            if ("type" in item) {
              await permissions.refreshPermissions({ item });
            } else {
              await permissions.refreshPermissions({ root: true });
            }

            timeouts[id]();
          },
          trigger
        );
      },
      resolveAndCommitData: async () => {
        const { config, state, dispatch, resolveComponentData } = get();

        walkTree(
          state,
          config,
          (content) => content,
          (childItem) => {
            resolveComponentData(childItem, "load").then((resolved) => {
              const { state } = get();

              const node = state.indexes.nodes[resolved.node.props.id];

              // Ensure node hasn't been deleted whilst resolution happens
              if (node && resolved.didChange) {
                if (resolved.node.props.id === "root") {
                  dispatch({
                    type: "replaceRoot",
                    root: toRoot(resolved.node),
                  });
                } else {
                  // Use latest position, in case it's moved
                  const zoneCompound = `${node.parentId}:${node.zone}`;
                  const parentZone = state.indexes.zones[zoneCompound];

                  const index = parentZone.contentIds.indexOf(
                    resolved.node.props.id
                  );

                  dispatch({
                    type: "replace",
                    data: resolved.node,
                    destinationIndex: index,
                    destinationZone: zoneCompound,
                  });
                }
              }
            });

            return childItem;
          }
        );
      },
    }))
  );

export const appStoreContext = createContext(createAppStore());

export function useAppStore<T>(selector: (state: AppStore) => T) {
  const context = useContext(appStoreContext);

  return useStore(context, selector);
}

export function useAppStoreApi() {
  return useContext(appStoreContext);
}
