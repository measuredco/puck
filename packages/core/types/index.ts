export * from "./API";
export * from "./API/Overrides";
export * from "./AppState";
export * from "./Config";
export * from "./Data";
export * from "./Fields";
export * from "./Props";
export * from "./Utils";
