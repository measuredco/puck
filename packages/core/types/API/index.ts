import { PuckAction } from "../../reducer";
import { AppState } from "./../AppState";
import { Data } from "./../Data";
import { Overrides } from "./Overrides";

export type Permissions = {
  drag: boolean;
  duplicate: boolean;
  delete: boolean;
  edit: boolean;
  insert: boolean;
} & Record<string, boolean>;

export type IframeConfig = {
  enabled?: boolean;
};

export type OnAction<UserData extends Data = Data> = (
  action: PuckAction,
  appState: AppState<UserData>,
  prevAppState: AppState<UserData>
) => void;

export type Plugin = {
  overrides: Partial<Overrides>;
};

export * from "./Viewports";

export type { Overrides };
