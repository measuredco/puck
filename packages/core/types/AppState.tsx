import { ItemSelector } from "../lib/get-item";
import { Viewport } from "./API";
import { Data } from "./Data";

export type ItemWithId = {
  _arrayId: string;
  _originalIndex: number;
};

export type ArrayState = { items: ItemWithId[]; openId: string };

export type UiState = {
  leftSideBarVisible: boolean;
  rightSideBarVisible: boolean;
  itemSelector: ItemSelector | null;
  arrayState: Record<string, ArrayState | undefined>;
  componentList: Record<
    string,
    {
      components?: string[];
      title?: string;
      visible?: boolean;
      expanded?: boolean;
    }
  >;
  isDragging: boolean;
  viewports: {
    current: {
      width: number;
      height: number | "auto";
    };
    controlsVisible: boolean;
    options: Viewport[];
  };
};

export type AppState<UserData extends Data = Data> = {
  data: UserData;
  ui: UiState;
};
