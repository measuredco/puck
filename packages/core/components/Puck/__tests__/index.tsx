import { act, render, screen } from "@testing-library/react";
import { Config } from "../../../types";
import "@testing-library/jest-dom";

jest.mock("../styles.module.css");
jest.mock("@dnd-kit/react");

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: (query: string) => ({
    matches: false, // default → desktop
    media: query,
    onchange: null,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    addListener: jest.fn(), // ⬅️ legacy APIs some libs still call
    removeListener: jest.fn(),
    dispatchEvent: jest.fn(),
  }),
});

jest.mock("@dnd-kit/react", () => {
  const original = jest.requireActual("@dnd-kit/react");
  return {
    ...original,
    // Provider becomes a no-op wrapper
    DragDropProvider: ({ children }: { children: React.ReactNode }) => (
      <>{children}</>
    ),

    // Hooks return dummy objects so destructuring works
    useDroppable: () => ({
      ref: () => undefined,
      setNodeRef: () => undefined,
      isOver: false,
    }),
    useDraggable: () => ({
      attributes: {},
      listeners: {},
      setNodeRef: () => undefined,
      isDragging: false,
    }),
  };
});

class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}
(global as any).ResizeObserver = ResizeObserver;

type PuckInternal = {
  appStore: AppStoreApi;
};

const getInternal = () => {
  return (window as any).__PUCK_INTERNAL_DO_NOT_USE as PuckInternal;
};

import { Puck } from "../index";
import { AppStoreApi } from "../../../store";

describe("Puck", () => {
  const componentARender = jest.fn(() => null);
  const componentBRender = jest.fn(() => null);
  const rootRender = jest.fn(() => null);

  const config: Config = {
    root: {
      render: ({ children }) => {
        rootRender();
        return <div>Root{children}</div>;
      },
    },
    components: {
      componentA: {
        render: () => {
          componentARender();
          return <div>Component A</div>;
        },
      },
      componentB: {
        render: () => {
          componentBRender();
          return <div>Component A</div>;
        },
      },
    },
  };

  afterEach(() => {
    rootRender.mockClear();
    componentARender.mockClear();
    componentBRender.mockClear();
  });

  // flush any queued state updates
  const flush = () => act(async () => {});

  it("root renders", async () => {
    render(<Puck config={config} data={{}} iframe={{ enabled: false }} />);

    await flush();

    expect(rootRender).toHaveBeenCalled();
    expect(screen.getByText("Root")).toBeInTheDocument();
  });

  it("should generate the correct state on mount", async () => {
    render(<Puck config={config} data={{}} iframe={{ enabled: false }} />);

    await flush();

    const { appStore } = getInternal();

    expect(appStore.getState()).toMatchSnapshot();
  });
});
