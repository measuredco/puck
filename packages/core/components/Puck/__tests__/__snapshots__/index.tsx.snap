// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puck should generate the correct state on mount 1`] = `
{
  "componentState": {},
  "config": {
    "components": {
      "componentA": {
        "render": [Function],
      },
      "componentB": {
        "render": [Function],
      },
    },
    "root": {
      "render": [Function],
    },
  },
  "dispatch": [Function],
  "fields": {
    "fields": {
      "title": {
        "type": "text",
      },
    },
    "id": undefined,
    "lastResolvedData": {},
    "loading": false,
  },
  "getComponentConfig": [Function],
  "history": {
    "back": [Function],
    "currentHistory": [Function],
    "forward": [Function],
    "hasFuture": [Function],
    "hasPast": [Function],
    "histories": [
      {
        "state": {
          "data": {
            "content": [],
            "root": {
              "props": {},
            },
            "zones": {},
          },
          "indexes": {
            "nodes": {
              "root": {
                "data": {
                  "props": {
                    "id": "root",
                  },
                  "type": "root",
                },
                "flatData": {
                  "props": {
                    "id": "root",
                  },
                  "type": "root",
                },
                "parentId": null,
                "path": [],
                "zone": "",
              },
            },
            "zones": {
              "root:default-zone": {
                "contentIds": [],
                "type": "root",
              },
            },
          },
          "ui": {
            "arrayState": {},
            "componentList": {},
            "field": {
              "focus": null,
            },
            "isDragging": false,
            "itemSelector": null,
            "leftSideBarVisible": true,
            "previewMode": "edit",
            "rightSideBarVisible": true,
            "viewports": {
              "controlsVisible": true,
              "current": {
                "height": "auto",
                "width": 360,
              },
              "options": [],
            },
          },
        },
      },
    ],
    "index": 0,
    "initialAppState": {
      "data": {
        "content": [],
        "root": {
          "props": {},
        },
        "zones": {},
      },
      "indexes": {
        "nodes": {
          "root": {
            "data": {
              "props": {
                "id": "root",
              },
              "type": "root",
            },
            "flatData": {
              "props": {
                "id": "root",
              },
              "type": "root",
            },
            "parentId": null,
            "path": [],
            "zone": "",
          },
        },
        "zones": {
          "root:default-zone": {
            "contentIds": [],
            "type": "root",
          },
        },
      },
      "ui": {
        "arrayState": {},
        "componentList": {},
        "field": {
          "focus": null,
        },
        "isDragging": false,
        "itemSelector": null,
        "leftSideBarVisible": true,
        "previewMode": "edit",
        "rightSideBarVisible": true,
        "viewports": {
          "controlsVisible": true,
          "current": {
            "height": "auto",
            "width": 360,
          },
          "options": [],
        },
      },
    },
    "nextHistory": [Function],
    "prevHistory": [Function],
    "record": [Function],
    "setHistories": [Function],
    "setHistoryIndex": [Function],
  },
  "iframe": {
    "enabled": false,
    "waitForStyles": true,
  },
  "metadata": undefined,
  "nodes": {
    "nodes": {},
    "registerNode": [Function],
    "unregisterNode": [Function],
  },
  "onAction": undefined,
  "overrides": {},
  "pendingLoadTimeouts": {},
  "permissions": {
    "cache": {},
    "getPermissions": [Function],
    "globalPermissions": {
      "delete": true,
      "drag": true,
      "duplicate": true,
      "edit": true,
      "insert": true,
    },
    "refreshPermissions": [Function],
    "resolvePermissions": [Function],
    "resolvedPermissions": {},
  },
  "plugins": [],
  "resolveAndCommitData": [Function],
  "resolveComponentData": [Function],
  "selectedItem": null,
  "setComponentLoading": [Function],
  "setComponentState": [Function],
  "setStatus": [Function],
  "setUi": [Function],
  "setZoomConfig": [Function],
  "state": {
    "data": {
      "content": [],
      "root": {
        "props": {},
      },
      "zones": {},
    },
    "indexes": {
      "nodes": {
        "root": {
          "data": {
            "props": {
              "id": "root",
            },
            "type": "root",
          },
          "flatData": {
            "props": {
              "id": "root",
            },
            "type": "root",
          },
          "parentId": null,
          "path": [],
          "zone": "",
        },
      },
      "zones": {
        "root:default-zone": {
          "contentIds": [],
          "type": "root",
        },
      },
    },
    "ui": {
      "arrayState": {},
      "componentList": {},
      "field": {
        "focus": null,
      },
      "isDragging": false,
      "itemSelector": null,
      "leftSideBarVisible": false,
      "previewMode": "edit",
      "rightSideBarVisible": false,
      "viewports": {
        "controlsVisible": true,
        "current": {
          "height": "auto",
          "width": 360,
        },
        "options": [],
      },
    },
  },
  "status": "READY",
  "unsetComponentLoading": [Function],
  "viewports": [
    {
      "height": "auto",
      "icon": "Smartphone",
      "label": "Small",
      "width": 360,
    },
    {
      "height": "auto",
      "icon": "Tablet",
      "label": "Medium",
      "width": 768,
    },
    {
      "height": "auto",
      "icon": "Monitor",
      "label": "Large",
      "width": 1280,
    },
  ],
  "zoomConfig": {
    "autoZoom": NaN,
    "rootHeight": 0,
    "zoom": NaN,
  },
}
`;
